### PR TITLE
Adding information how to control randomness with DataLoader

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45749 Adding information how to control randomness with DataLoader**

Fixes: #41329
Fixes: #38541

Differential Revision: [D24088407](https://our.internmc.facebook.com/intern/diff/D24088407)

### CODE REVIEW

File: docs/source/notes/randomness.rst [hunk 1]
```diff
@@ -121,3 +123,38 @@ CUDA RNN and LSTM
 In some versions of CUDA, RNNs and LSTM networks may have non-deterministic behavior.
 See :meth:`torch.nn.RNN` and :meth:`torch.nn.LSTM` for details and workarounds.
 
+CUDA random seed
+----------------
+
+Also it is recommended to set CUDAs random seed::
+
+    torch.cuda.manual_seed_all(0)
```
Comment (albanD) [inline docs/source/notes/randomness.rst:131]: `torch.manual_seed()` already sets this no?

File: docs/source/notes/randomness.rst [hunk 2]
```diff
@@ -121,3 +123,38 @@ CUDA RNN and LSTM
...
+    torch.cuda.manual_seed(0)
+
+Python
```
Comment (albanD) [inline docs/source/notes/randomness.rst:134]: This should be inside "controlling sources of randomness" above no?

File: docs/source/notes/randomness.rst [hunk 3]
```diff
@@ -121,3 +123,38 @@ CUDA RNN and LSTM
...
+
+DataLoader
+..........
+
+DataLoader will reseed workers following :ref:`data-loading-randomness` algorithm.
+Use :meth:`worker_init_fn` to preserve reproducibility::
+
+    def seed_worker(worker_id):
+        worker_seed = torch.initial_seed() % 2**32
+        numpy.random.seed(worker_seed)
```
Comment (albanD) [inline docs/source/notes/randomness.rst:150]: Maybe add a comment here on why there is no `torch.manual_seed()` ?

File: docs/source/notes/randomness.rst (hunk 1 above)
Comment (VitalyFedyunin) [reply docs/source/notes/randomness.rst:131]: Ah, right, just verified inside of the code.

File: docs/source/notes/randomness.rst (hunk 2 above)
Comment (VitalyFedyunin) [reply docs/source/notes/randomness.rst:134]: Agree, moved

File: docs/source/notes/randomness.rst (hunk 3 above)
Comment (VitalyFedyunin) [reply docs/source/notes/randomness.rst:150]: I think in some cases users might want to use torch.manual_seed(), lets just leave it as is so they have idea where to look, and they can elaborate solution later on.

**Review by albanD — APPROVED** [review]
LGTM

File: torch/utils/data/dataloader.py [hunk 4]
```diff
@@ -139,6 +139,9 @@ class DataLoader(Generic[T_co]):
                  See `Dataset Types`_ for more details on these two types of datasets and how
                  :class:`~torch.utils.data.IterableDataset` interacts with
                  `Multi-process data loading`_.
+
+    .. warning:: See :ref:`reproducibility`, and :ref:`dataloader-workers-random-seed`, and
```
Comment (mruberry) [inline torch/utils/data/dataloader.py:143]: This seems more like a note:: than a warning:: to me.

It's great to link these docs here. Can this be more descriptive than "random seed related questions?" Maybe something like:

"See the x, y, and z notes for information about reproducible data loading."?

File: docs/source/notes/randomness.rst [hunk 5]
```diff
@@ -121,3 +131,22 @@ CUDA RNN and LSTM
 In some versions of CUDA, RNNs and LSTM networks may have non-deterministic behavior.
 See :meth:`torch.nn.RNN` and :meth:`torch.nn.LSTM` for details and workarounds.
 
+DataLoader
+..........
+
+DataLoader will reseed workers following :ref:`data-loading-randomness` algorithm.
+Use :meth:`worker_init_fn` to preserve reproducibility::
```
Comment (mruberry) [inline docs/source/notes/randomness.rst:138]: "preserve reproducibility" -> "to set the seed(s) on each worker"

Would that be more direct? Should this mention why the worker_seed is computed as it is?

File: docs/source/notes/randomness.rst [hunk 6]
```diff
@@ -121,3 +131,22 @@ CUDA RNN and LSTM
 In some versions of CUDA, RNNs and LSTM networks may have non-deterministic behavior.
 See :meth:`torch.nn.RNN` and :meth:`torch.nn.LSTM` for details and workarounds.
 
+DataLoader
+..........
+
+DataLoader will reseed workers following :ref:`data-loading-randomness` algorithm.
```
Comment (mruberry) [inline docs/source/notes/randomness.rst:137]: "PyTorch's dataloaders will reseed workers using the ..."

File: docs/source/notes/randomness.rst [hunk 7]
```diff
@@ -30,6 +32,14 @@ CPU and CUDA)::
     import torch
     torch.manual_seed(0)
 
+Python
+------
+
+For custom operators, you might need to set python seed as well::
```
Comment (mruberry) [inline docs/source/notes/randomness.rst:38]: "to seed Python's random number generator"

**Review by mruberry — APPROVED** [review]
Great to see these docs connected to address this point of user confusion.

I made a few suggestions.
